pulsar-transaction/coordinator/src/test/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriterTest.java

```diff
@@ -0,0 +1,331 @@
+        Assert.assertEquals(entry.getDataBuffer().readInt(), 1);
+        entry.release();
+        // cleanup.
+        txnLogBufferedWriter.close();
+        managedLedger.close();
+    }
+
+    /**
+     * Adjustable thresholds: trigger BookKeeper-write when reaching any one of the following conditions
+     *     Max size (bytes)
+     *     Max records count
+     *     Max delay time
+     * Tests these three thresholds.
+     */
+    @Test
+    public void testFlushThresholds() throws Exception{
+        // Create components.
+        ManagedLedger managedLedger = Mockito.mock(ManagedLedger.class);
+        Mockito.when(managedLedger.getName()).thenReturn("-");
+        OrderedExecutor orderedExecutor =  OrderedExecutor.newBuilder()
+                .numThreads(5).name("tx-brokers-topic-workers").build();
+        ScheduledExecutorService scheduledExecutorService =
+                Executors.newSingleThreadScheduledExecutor(new DefaultThreadFactory("pulsar-stats-updater"));
+        TxnLogBufferedWriter.DataSerializer<Integer> serializer = new TxnLogBufferedWriter.DataSerializer<Integer>(){
+            @Override
+            public int getSerializedSize(Integer data) {
+                return 4;
+            }
+            @Override
+            public ByteBuf serialize(Integer data) {
+                return null;
+            }
+            @Override
+            public ByteBuf serialize(ArrayList<Integer> dataArray) {
+                int sum = CollectionUtils.isEmpty(dataArray) ? 0 : dataArray.stream().reduce((a, b) -> a+b).get();
+                ByteBuf byteBuf = Unpooled.buffer(4);
+                byteBuf.writeInt(sum);
+                return byteBuf;
+            }
+        };
+        List<Integer> flushedDataList = new ArrayList<>();
+        Mockito.doAnswer(new Answer() {
+            @Override
+            public Object answer(InvocationOnMock invocation) throws Throwable {
+                ByteBuf byteBuf = (ByteBuf)invocation.getArguments()[0];
+                byteBuf.skipBytes(4);
+                flushedDataList.add(byteBuf.readInt());
+                AsyncCallbacks.AddEntryCallback callback =
+                        (AsyncCallbacks.AddEntryCallback) invocation.getArguments()[1];
+                callback.addComplete(PositionImpl.get(1,1), byteBuf,
+                        invocation.getArguments()[2]);
+                return null;
+            }
+        }).when(managedLedger).asyncAddEntry(Mockito.any(ByteBuf.class), Mockito.any(), Mockito.any());
+
+        TxnLogBufferedWriter txnLogBufferedWriter = new TxnLogBufferedWriter<>(managedLedger, orderedExecutor,
+                scheduledExecutorService, serializer, 32, 1024 * 4, 100, true);
+        TxnLogBufferedWriter.AddDataCallback callback = Mockito.mock(TxnLogBufferedWriter.AddDataCallback.class);
+        // Test threshold: writeMaxDelayInMillis.
+        txnLogBufferedWriter.asyncAddData(100, callback, 100);
```

Comment:
I don't understand why the ctx here is 100, does it only play a role in the log?

Reply:
No, in the current case: the param-context argument is meaningless; it can be any value.